libflux/flux-core/src/semantic/tests/labels.rs

```diff
@@ -0,0 +1,226 @@
+    }
+}
+
+#[test]
+fn merge_labels_to_string() {
+    test_infer! {
+        config: AnalyzerConfig{
+            features: vec![Feature::LabelPolymorphism],
+            ..AnalyzerConfig::default()
+        },
+        src: r#"
+            x = if 1 == 1 then "a" else "b"
+            y = if 1 == 1 then "a" else "b" + "b"
+            z = ["a", "b"]
+        "#,
+        exp: map![
+            "x" => "string",
+            "y" => "string",
+            "z" => "[string]",
+        ],
+    }
+}
+
+#[test]
+fn merge_labels_to_string_in_function() {
+    test_infer! {
+        config: AnalyzerConfig{
+            features: vec![Feature::LabelPolymorphism],
+            ..AnalyzerConfig::default()
+        },
+        env: map![
+            "same" => "(x: A, y: A) => A"
+        ],
+        src: r#"
+            x = same(x: "a", y: "b")
+            y = same(x: ["a"], y: ["b"])
+        "#,
+        exp: map![
+            "x" => "string",
+            "y" => "[string]",
+        ],
+    }
+}
+
+#[test]
+fn attempt_to_use_label_polymorphism_without_feature() {
+    test_error_msg! {
+        env: map![
+            "columns" => "(table: A, ?column: C) => { C: string } where A: Record, C: Label",
+        ],
+        src: r#"
+            x = columns(table: { a: 1, b: "b" }, column: "abc")
+            y = x.abc
+        "#,
+        expect: expect![[r#"
+            error: string is not Label (argument column)
+              ┌─ main:2:58
+              │
+            2 │             x = columns(table: { a: 1, b: "b" }, column: "abc")
+              │                                                          ^^^^^
+
+            error: record is missing label abc
+              ┌─ main:3:17
+              │
+            3 │             y = x.abc
+              │                 ^
+
+        "#]],
+    }
+}
+#[test]
+fn columns() {
+    test_infer! {
+        config: AnalyzerConfig{
+            features: vec![Feature::LabelPolymorphism],
+            ..AnalyzerConfig::default()
+        },
+        env: map![
+            "stream" => "stream[{ a: int }]",
+            "columns" => "(<-tables: stream[A], ?column: C) => stream[{ C: string }] where A: Record, C: Label",
+            "map" => "(<-tables: stream[A], fn: (r: A) => B) => stream[B]"
+        ],
+        src: r#"
+            x = stream
+                |> columns(column: "abc")
+                |> map(fn: (r) => ({ x: r.abc }))
+        "#,
+        exp: map![
+            "x" => "stream[{ x: string }]",
+        ],
+    }
+}
+
+#[test]
+fn optional_label() {
+    test_error_msg! {
+        config: AnalyzerConfig{
+            features: vec![Feature::LabelPolymorphism],
+            ..AnalyzerConfig::default()
+        },
+        env: map![
+            "columns" => "(table: A, ?column: C) => { C: string } where A: Record, C: Label",
+        ],
+        src: r#"
+            x = columns(table: { a: 1, b: "b" })
+            y = x.abc
+        "#,
+        // TODO Improve this error
```

Comment:
Is this still a TODO? The error below is a bit confusing. I would have expected an error about `abc` not existing on `x`.

Reply:
It is due to `Property` being tweaked https://github.com/influxdata/flux/pull/4388/files#diff-d5ab296ff16df9447d31cf953c8c275a43922b7cd3b260af30e0325abbdd53eaR1886-R1889 . Since keys may contain variables they need to be `Substitutable` which forces `String` to be `Substitutable`.